### ISSUE TITLE
NullPointerException in AlarmPingSender

### ISSUE THREAD

**Issue by noordawod** [eclipse/paho.mqtt.android#158]
Please fill out the form below before submitting, thank you!

- [X] Bug exists Release Version 1.1.0 (Java Repository Master Branch)
- [ ] Bug exists in Snapshot Version 1.1.1-SNAPSHOT (Android Service Repository Master Branch)
- [ ] Bug is just in the Sample Application.

__Android API Version Bug Seen on:__

__Android Version Bug Seen on:__

Android 7.1

Please also check that if you have found the bug in the Release version (1.1.0) that you check that it also exists in the Snapshot (1.1.1-SNAPSHOT) before raising a bug.

## Description of Bug:

I tracked down the bug to `AlarmPingSender.stop()` calling `AlarmManager.cancel()` with a null pending intent. I looked at SNAPSHOT-1.1.1 and saw there's a new `if()` that checks against a boolean, but still no check whether the intent is null.

https://github.com/eclipse/paho.mqtt.android/blob/master/org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/AlarmPingSender.java#L87

I couldn't find another call to `cancel()` in the source code (luckily) so this is an easy fix.

## Console Log output (if available):

Trace from Play console for a published app:

```
java.lang.NullPointerException: cancel() called with a null PendingIntent
	at android.app.AlarmManager.cancel(AlarmManager.java:890)
	at org.eclipse.paho.android.service.a.b(Unknown Source)
	at org.eclipse.paho.client.mqttv3.a.a.a(Unknown Source)
	at org.eclipse.paho.client.mqttv3.a.a$a.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:761)
```

**Comment by elsennov** [comment]
When is the release to fix this @jpwsutton ? Thanks!

**Comment by brstgt** [comment]
It looks like this always happens if an exception is thrown during connection (in ConnectBG.run). AlarmPingSender.pendingIntent is initialized in AlarmPingSender.start. This is called only on successfull connection start which is quite reasonable as pings are only needed for an established connection. But when the connection is shut down BEFORE AlarmPingSender is started then it should also cancel the intent only when it has been created (and is not null)

**Comment by piyushgupta27** [comment]
Hey, Has this fix been released on the maven repo? The last update to the Snapshot builds (https://repo.eclipse.org/content/repositories/paho-snapshots/org/eclipse/paho/) was Nov 9. Can you please let me know is there a way to get the fix without having to fork the repo?

**Comment by vanshg** [comment]
@piyushgupta27 The 1.1.1 SNAPSHOT that I've been using has been working for me 

'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1-20160907.104533-5@aar'

**Comment by SameerMB** [comment]
@vanshg I'm getting gradle error in Android studio
Error:Could not find org.eclipse.paho:org.eclipse.paho.android.service:1.1.1-20160907.104533-5.
Required by:
    android-app:app:unspecified

**Comment by vanshg** [comment]
@SameerMB you need to use the snapshot repo

**Comment by SameerMB** [comment]
Thanks @vanshg , managed to build it. Let me try if it still crashes.

**Comment by SameerMB** [comment]
It's working, thanks @vanshg

**Comment by brstgt** [comment]
Wenn will an official release with this fix be available? Working with snapshots for several months is "not so cool".

**Comment by jpwsutton** [comment]
The next version of Paho will be 1.3.0 which will simultaneously release with Eclipse Oxygen. This will be some time between now and June 28th.

**Comment by brstgt** [comment]
Why does a bugfix release has to wait for a big bang release of an epic project?

**Comment by elsennov** [comment]
Agree with @brstgt , just why?

**Comment by jpwsutton** [comment]
Not a problem, if you would like a bug fix build I can run one. I'd just left it as nobody had asked for one yet. I'll kick the process off now and let you know once 1.1.1 is available on the master repository.

**Comment by jpwsutton** [comment]
Bugfix Release 1.1.1 is now available here: https://repo.eclipse.org/content/repositories/paho-releases/org/eclipse/paho/org.eclipse.paho.android.service/1.1.1/  let me know if there are any issues with it.

**Comment by vanshg** [comment]
Awesome, thanks @jpwsutton! Is there any chance this can be hosted on jCenter too? Or are the builds only going to be hosted on eclipse's own servers?

**Comment by RahulSanscrityayan** [comment]
I am using 

compile 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.0.3-SNAPSHOT'
    compile 'org.eclipse.paho:org.eclipse.paho.android.service:1.0.3-SNAPSHOT'

in android and getting 


java.lang.NullPointerException: cancel() called with a null PendingIntent
                                                                      at android.app.AlarmManager.cancel(AlarmManager.java:951)
                                                                      at org.eclipse.paho.android.service.AlarmPingSender.stop(AlarmPingSender.java:87)
                                                                      at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:353)
                                                                      at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:668)
                                                                      at java.lang.Thread.run(Thread.java:761)

**Comment by jpwsutton** [comment]
@RahulSanscrityayan Looks like you're using the 1.0.3 snapshot, as mentioned earlier this was fixed in the 1.1.1 release available on Maven Central and the Eclipse Repositories.

**Comment by gourav1908** [comment]
> 1.1.1-20160907.104533-5

@SameerMB Could you please let me know how can I use this particular snapshop repo, I am stuck from 4 days. TIA
I need to make it work in Target 33
